syntax = "proto3";

package chat;

// The ChatService defines all RPC methods.
service ChatService {
  rpc CreateAccount(CreateAccountRequest) returns (CreateAccountResponse);
  rpc Login(LoginRequest) returns (LoginResponse);
  rpc Logout(LogoutRequest) returns (LogoutResponse);
  rpc ListAccounts(ListAccountsRequest) returns (ListAccountsResponse);
  rpc SendMessage(SendMessageRequest) returns (SendMessageResponse);
  rpc ReadMessages(ReadMessagesRequest) returns (ReadMessagesResponse);
  rpc ListMessages(ListMessagesRequest) returns (ListMessagesResponse);
  rpc DeleteMessages(DeleteMessagesRequest) returns (DeleteMessagesResponse);
  rpc DeleteAccount(DeleteAccountRequest) returns (DeleteAccountResponse);
  rpc SubscribeMessages(SubscribeRequest) returns (stream ChatMessage);
}

message CreateAccountRequest {
  string username = 1;
  string password = 2;
}

message CreateAccountResponse {
  bool success = 1;
  string message = 2;
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message LoginResponse {
  bool success = 1;
  string message = 2;
  int32 unread_messages = 3;
}

message LogoutRequest {
  string username = 1;
}

message LogoutResponse {
  bool success = 1;
  string message = 2;
}

message ListAccountsRequest {
  string pattern = 1;
  int32 page = 2;
}

message ListAccountsResponse {
  repeated string accounts = 1;
}

message SendMessageRequest {
  string sender = 1;
  string to = 2;
  string message = 3;
}

message SendMessageResponse {
  bool success = 1;
  string message = 2;
}

message ChatMessage {
  int64 id = 1;
  string sender = 2;
  string to = 3;
  string message = 4;
  int64 timestamp = 5;
}

message ReadMessagesRequest {
  string username = 1;
  int32 count = 2;
}

message ReadMessagesResponse {
  repeated ChatMessage messages = 1;
}

message ListMessagesRequest {
  string username = 1;
}

message ListMessagesResponse {
  repeated ChatMessage messages = 1;
}

message DeleteMessagesRequest {
  string username = 1;
  repeated int64 message_ids = 2;
}

message DeleteMessagesResponse {
  bool success = 1;
  string message = 2;
}

message DeleteAccountRequest {
  string username = 1;
}

message DeleteAccountResponse {
  bool success = 1;
  string message = 2;
}

// Request message for subscribing to messages.
message SubscribeRequest {
  string username = 1;
}
